src/bse/K_stored_in_a_nest_matrix.F

```diff
@@ -0,0 +1,337 @@
+!
+!        Copyright (C) 2000-2023 the YAMBO team
+!              http://www.yambo-code.org
+!
+! Authors (see AUTHORS file for details): 
+! 
+! This file is distributed under the terms of the GNU 
+! General Public License. You can redistribute it and/or 
+! modify it under the terms of the GNU General Public 
+! License as published by the Free Software Foundation; 
+! either version 2, or (at your option) any later version.
+!
+! This program is distributed in the hope that it will 
+! be useful, but WITHOUT ANY WARRANTY; without even the 
+! implied warranty of MERCHANTABILITY or FITNESS FOR A 
+! PARTICULAR PURPOSE.  See the GNU General Public License 
+! for more details.
+!
+! You should have received a copy of the GNU General Public 
+! License along with this program; if not, write to the Free 
+! Software Foundation, Inc., 59 Temple Place - Suite 330,Boston, 
+! MA 02111-1307, USA or visit http://www.gnu.org/copyleft/gpl.txt.
+!
+
+module shellmat_module
+#include <petsc/finclude/petscmat.h>
+  use petscmat
+  Mat::R,C
+end module shellmat_module
+
+subroutine ignore_petsc_error_handler(comm,line,fun,file,n,p,mess,ctx,ierr)
+!
+! Handler to ignore error if MatMultHermitianTranspose cannot be set.
+! This operation is only available for shell matrices in PETSc from v.3.21
+! Setting MatMultTranspose is enough, but it is preferable to set both.
+!
+  use petscsysdef
+  integer line,n,p
+  PetscInt ctx 
+  PetscErrorCode ierr
+  MPI_Comm comm
+  character*(*) fun,file,mess
+end subroutine ignore_petsc_error_handler
+
+subroutine K_stored_in_a_nest_matrix(i_BS_mat,slepc_mat)
+!
+! BS_K_coupling and l_BS_ares_from_res are assumed TRUE
+!
+! K is stored in a PETSc nest matrix, taking advantage of the four block
+! structure of the matrix in this case:
+!
+! | R     C  |
+! |-C^*  -R^T|
+!
+#include <petsc/finclude/petscsys.h>
+#include <petsc/finclude/petscvec.h>
+#include <petsc/finclude/petscmat.h>
+#include <petsc/finclude/petscvec.h>
+#include <slepc/finclude/slepcsys.h>
+#include <slepc/finclude/slepceps.h>
+!
+  use pars,           ONLY:cI,cONE
+  use BS,             ONLY:BS_K_dim,BS_H_dim,BS_blk,n_BS_blks
+  use BS_solvers,     ONLY:BSS_eh_E,BSS_eh_W,BSS_perturbative_width
+  use cuda_m,         ONLY:have_cuda
+  !
+  use petscmat
+  use slepceps
+  use slepcepsdef
```

Comment:
Should not be necessary, as `slepceps` already includes the definitions module `slepcepsdef`.
```suggestion
```